scripts/terminalserver/trees.jl

```diff
@@ -0,0 +1,156 @@
+struct LazyTree
+    head::String
+    type::String
+    children
+end
+
+struct SubTree
+    head::String
+    child
+end
+
+LazyTree(head, children) = LazyTree(head, "", children)
+
+struct Leaf
+    val
+end
+
+const TREES = Dict{Int, LazyTree}()
+const ID = Ref(0)
+
+treeid() = (ID[] += 1)
+
+function treerender(x::LazyTree)
+    id = treeid()
+    TREES[id] = x
+
+    return Dict(
+        :head => x.head,
+        :id => id,
+        :haschildren => true,
+        :lazy => true,
+        :value => ""
+    )
+end
+
+function treerender(x::SubTree)
+    child = treerender(x.child)
+
+    return Dict(
+        :head => x.head,
+        :value => get(child, :head, ""),
+        :haschildren => get(child, :haschildren, true),
+        :id => get(child, :id, -1),
+        :lazy => get(child, :lazy, true)
+    )
+end
+
+function treerender(x::Leaf)
+    return Dict(
+        :head => x.val,
+        :id => -1,
+        :value => "",
+        :haschildren => false,
+        :lazy => false
+    )
+end
+
+getfield_safe(x, f, default = "#undef") = isdefined(x, f) ? getfield(x, f) : default
+
+function treerender(x)
+    fields = fieldnames(typeof(x))
+
+    if isempty(fields)
+        treerender(Text(string(typeof(x), "()")))
+    else
+        treerender(LazyTree(string(typeof(x)), function ()
+            [SubTree(string(f), getfield_safe(x, f)) for f in fields]
+        end))
+    end
+end
+
+function treerender(x::AbstractDict)
+    treerender(LazyTree(string(nameof(typeof(x)), "{$(eltype(x).parameters[1]), $(eltype(x).parameters[2])}"), function ()
+        if length(keys(x)) > 25
+            partition_by_keys(x)
+        else
+            [SubTree(repr(k), x[k]) for k in keys(x)]
+        end
+    end))
+end
+
+function treerender(x::Module)
+    treerender(LazyTree(string(x), function ()
+        ns = names(x, all = true)
+        out = []
+        for n in ns
+            isdefined(x, n) || continue
+            Base.isdeprecated(x, n) && continue
+            startswith(string(n), '#') && continue
+            push!(out, SubTree(string(n), getfield(x, n)))
+        end
+
+        out
+    end))
+end
```

Comment:
maybe we want to filter out `x` module itself here ?

Reply:
Good idea. Done.